File: src/frontend/src/lib/api-models/permissions.ts
import type {
  OrgPermissions as ApiOrgPermissions,
  ProjectPermissions as ApiProjectPermissions,
} from '@ssn/backend-api';

export type OrgPermissions = {
  orgAdmin: boolean;
  memberManage: boolean;
  teamManage: boolean;
  projectCreate: boolean;
  billingManage: boolean;
};

export type ProjectPermissions = {
  projectAdmin: boolean;
  canisterManage: boolean;
  proposalCreate: boolean;
  proposalApprove: boolean;
  canisterOperate: boolean;
  canisterRead: boolean;
  approvalPolicyManage: boolean;
  projectSettings: boolean;
};

export const emptyOrgPermissions: OrgPermissions = {
  orgAdmin: false,
  memberManage: false,
  teamManage: false,
  projectCreate: false,
  billingManage: false,
};

export const emptyProjectPermissions: ProjectPermissions = {
  projectAdmin: false,
  canisterManage: false,
  proposalCreate: false,
  proposalApprove: false,
  canisterOperate: false,
  canisterRead: false,
  approvalPolicyManage: false,
  projectSettings: false,
};

export function mapOrgPermissions(res: ApiOrgPermissions): OrgPermissions {
  return {
    orgAdmin: res.org_admin,
    memberManage: res.member_manage,
    teamManage: res.team_manage,
    projectCreate: res.project_create,
    billingManage: res.billing_manage,
  };
}

export function mapOrgPermissionsToApi(
  perms: OrgPermissions,
): ApiOrgPermissions {
  return {
    org_admin: perms.orgAdmin,
    member_manage: perms.memberManage,
    team_manage: perms.teamManage,
    project_create: perms.projectCreate,
    billing_manage: perms.billingManage,
  };
}

export function mapProjectPermissions(
  res: ApiProjectPermissions,
): ProjectPermissions {
  return {
    projectAdmin: res.project_admin,
    canisterManage: res.canister_manage,
    proposalCreate: res.proposal_create,
    proposalApprove: res.proposal_approve,
    canisterOperate: res.canister_operate,
    canisterRead: res.canister_read,
    approvalPolicyManage: res.approval_policy_manage,
    projectSettings: res.project_settings,
  };
}

export function mapProjectPermissionsToApi(
  perms: ProjectPermissions,
): ApiProjectPermissions {
  return {
    project_admin: perms.projectAdmin,
    canister_manage: perms.canisterManage,
    proposal_create: perms.proposalCreate,
    proposal_approve: perms.proposalApprove,
    canister_operate: perms.canisterOperate,
    canister_read: perms.canisterRead,
    approval_policy_manage: perms.approvalPolicyManage,
    project_settings: perms.projectSettings,
  };
}

File: src/backend/src/mapping/permissions.rs
use crate::{data, dto};

pub fn map_org_permissions(perms: data::OrgPermissions) -> dto::OrgPermissions {
    dto::OrgPermissions {
        org_admin: perms.contains(data::OrgPermissions::ORG_ADMIN),
        member_manage: perms.contains(data::OrgPermissions::MEMBER_MANAGE),
        team_manage: perms.contains(data::OrgPermissions::TEAM_MANAGE),
        project_create: perms.contains(data::OrgPermissions::PROJECT_CREATE),
        billing_manage: perms.contains(data::OrgPermissions::BILLING_MANAGE),
    }
}

pub fn map_org_permissions_from_dto(perms: dto::OrgPermissions) -> data::OrgPermissions {
    let mut out = data::OrgPermissions::EMPTY;
    if perms.org_admin {
        out = out.union(data::OrgPermissions::ORG_ADMIN);
    }
    if perms.member_manage {
        out = out.union(data::OrgPermissions::MEMBER_MANAGE);
    }
    if perms.team_manage {
        out = out.union(data::OrgPermissions::TEAM_MANAGE);
    }
    if perms.project_create {
        out = out.union(data::OrgPermissions::PROJECT_CREATE);
    }
    if perms.billing_manage {
        out = out.union(data::OrgPermissions::BILLING_MANAGE);
    }
    out
}

pub fn map_project_permissions(perms: data::ProjectPermissions) -> dto::ProjectPermissions {
    dto::ProjectPermissions {
        project_admin: perms.contains(data::ProjectPermissions::PROJECT_ADMIN),
        canister_manage: perms.contains(data::ProjectPermissions::CANISTER_MANAGE),
        proposal_create: perms.contains(data::ProjectPermissions::PROPOSAL_CREATE),
        proposal_approve: perms.contains(data::ProjectPermissions::PROPOSAL_APPROVE),
        canister_operate: perms.contains(data::ProjectPermissions::CANISTER_OPERATE),
        canister_read: perms.contains(data::ProjectPermissions::CANISTER_READ),
        approval_policy_manage: perms.contains(data::ProjectPermissions::APPROVAL_POLICY_MANAGE),
        project_settings: perms.contains(data::ProjectPermissions::PROJECT_SETTINGS),
    }
}

pub fn map_project_permissions_from_dto(
    perms: dto::ProjectPermissions,
) -> data::ProjectPermissions {
    let mut out = data::ProjectPermissions::EMPTY;
    if perms.project_admin {
        out = out.union(data::ProjectPermissions::PROJECT_ADMIN);
    }
    if perms.canister_manage {
        out = out.union(data::ProjectPermissions::CANISTER_MANAGE);
    }
    if perms.proposal_create {
        out = out.union(data::ProjectPermissions::PROPOSAL_CREATE);
    }
    if perms.proposal_approve {
        out = out.union(data::ProjectPermissions::PROPOSAL_APPROVE);
    }
    if perms.canister_operate {
        out = out.union(data::ProjectPermissions::CANISTER_OPERATE);
    }
    if perms.canister_read {
        out = out.union(data::ProjectPermissions::CANISTER_READ);
    }
    if perms.approval_policy_manage {
        out = out.union(data::ProjectPermissions::APPROVAL_POLICY_MANAGE);
    }
    if perms.project_settings {
        out = out.union(data::ProjectPermissions::PROJECT_SETTINGS);
    }
    out
}
